### PR TITLE
Deprecate Custom Search Site Restricted JSON API

### PR DESCRIPTION
**Pull request by jonflynng**
## What does this change?

See issue for context: https://github.com/guardian/ophan/issues/5715

I chose to use the REST API over the Java client library for these reasons:

- The client library not compatible with API keys, only with Oauth 2.0 flow. This means a bit more work setting up service accounts and granting the right permissions to the right services.
- Given the lack of documentation, navigating the client library is time consuming and I think the trade off we get for the strong typing to just make one API call isn't worth it.

Rest docs: https://cloud.google.com/generative-ai-app-builder/docs/reference/rest

Sibling PR in other indexing repo [here](https://github.com/guardian/google-search-indexing-observatory/pull/60).

## Important

During testing, I discovered that approximately 15% of articles available in the old API and Google Search Results cannot be found in the new API. We'll monitor API request volume after this is deployed to track how frequently this occurs.

We intend to contact GCP support about these discrepancies between the old and new APIs, we have a limited relationship with them and it may take a while according to Mike Taylor. We've implemented a fallback solution: when the initial search fails, perform a second search using the article's webTitle instead of the CAPI ID suffix. This alternative method has successfully retrieved the missing articles. See images below.

**Example of article not found in new API:**

https://www.theguardian.com/australia-news/2024/nov/21/afternoon-update-thursday-ntwnfb

1. Article is correctly found (the top one) in Google search results when searching via CAPI ID suffix. This is the same behaviour as the old API:

Screenshot 2024-11-21 at 11 49 51

2. Article is not found in new API with CAPI id suffix:

Screenshot 2024-11-21 at 11 49 57

3. Article is found in new API using the `webTitle`:

Screenshot 2024-11-21 at 11 50 25

## How can we measure success?

Nothing changes.

## Accessibility

-   [ ] [Tested with screen reader](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#screen-reader)
-   [ ] [Navigable with keyboard](https://github.com/guardian/accessibility/blob/main/people-and-technology/02-physical.md#keyboard)
-   [ ] [Colour contrast passed](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#contrast)
-   [ ] [The change doesn't use only colour to convey meaning](https://github.com/guardian/accessibility/blob/main/people-and-technology/03-visual.md#use-of-colour)

### CODE REVIEW

**Review by sophie-macmillan — APPROVED** [review]

File: src/main/scala/ophan/google/index/checker/GoogleSearchService.scala [hunk 1]
```diff
@@ -1,56 +1,90 @@
...
+      def performSearch(searchTerm: String): Boolean = {
+        val requestBody = ujson.Obj(
+          "query" -> searchTerm,
+          "pageSize" -> 10
+        )
 
+        val request = HttpRequest.newBuilder()
+          .uri(URI.create(s"$baseUrl?key=$apiKey"))
+          .header("Content-Type", "application/json")
+          .header("User-Agent", "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36")
```
Comment (SHession) [inline src/main/scala/ophan/google/index/checker/GoogleSearchService.scala:59]: Why does this need to be set?

File: src/main/scala/ophan/google/index/checker/GoogleSearchService.scala [hunk 2]
```diff
@@ -1,56 +1,90 @@
...
+          .POST(HttpRequest.BodyPublishers.ofString(requestBody.toString()))
+          .build()
+
+        val response = httpClient.send(request, java.net.http.HttpResponse.BodyHandlers.ofString())
+        print(response.body())
```
Comment (SHession) [inline src/main/scala/ophan/google/index/checker/GoogleSearchService.scala:64]: Should this be removed?

**Review by SHession — COMMENTED** [review]
I think it's worth updating the documentation to provide some detail on the new libraries and services being used!

File: build.sbt [hunk 3]
```diff
@@ -23,10 +23,8 @@ libraryDependencies ++= Seq(
   "org.slf4j" % "log4j-over-slf4j" % "2.0.12", //  log4j-over-slf4j provides `org.apache.log4j.MDC`, which is dynamically loaded by the Lambda runtime
   "ch.qos.logback" % "logback-classic" % "1.5.3",
   "com.lihaoyi" %% "upickle" % "3.2.0",
-
   "com.madgag" %% "scala-collection-plus" % "0.11",
   "com.google.http-client" % "google-http-client-gson" % "1.44.1",
```
Comment (SHession) [inline build.sbt:27]: Can we also remove this other Google library?

File: src/main/scala/ophan/google/index/checker/GoogleSearchService.scala (hunk 2 above)
Comment (jonflynng) [reply src/main/scala/ophan/google/index/checker/GoogleSearchService.scala:64]: yep

File: src/main/scala/ophan/google/index/checker/GoogleSearchService.scala (hunk 1 above)
Comment (jonflynng) [reply src/main/scala/ophan/google/index/checker/GoogleSearchService.scala:59]: Good question, just copied it, ill have a look

Comment (jonflynng) [reply src/main/scala/ophan/google/index/checker/GoogleSearchService.scala:59]: This shouldn't need to be set, I will double check and test

File: build.sbt (hunk 3 above)
Comment (jonflynng) [reply build.sbt]: yes

**Pull request merged**